src/typeEvaluator/typeEvaluate.ts

```diff
@@ -960,12 +1037,37 @@ export function walk({node, scope}: {node: ExprNode; scope: Scope}): TypeNode {
       return handleOpCallNode(node, scope)
     }
 
-    case 'And':
+    case 'And': {
+      const left = walk({node: node.left, scope})
+      const right = walk({node: node.right, scope})
+      return mapUnion(left, scope, (lhs) =>
+        mapUnion(right, scope, (rhs) => {
+          return {
+            type: 'boolean',
+            value: nodeMatch(lhs) && nodeMatch(rhs),
+          } satisfies BooleanTypeNode
+        }),
+      )
+    }
```

Comment:
I think was slightly broken before (because it _always_ returned a `boolean`), but this seems a bit imprecise in how it deals with different types. `&&` and `||` always returns one of `true`, `false`, `null`. So for instance:

- If LHS and RHS are both `boolean` we know that this is always a `boolean`.
- However, if either LHS or RHS are `unknown` then this is actually `boolean | null`.
- Similarly, if LHS or RHS are _known_ to be either `object`, `number`, `array` then it's actually _always_ `null`.

These also quite interestingly compose a bit similarly under `union`.

I think maybe all of this is better expressed with something which turns a `TypeNode` into a type such as:

```ts
type BooleanInterpretation = {
  canBeTrue: boolean
  canBeFalse: boolean
  canBeNull: boolean
}
``` 

`canBeNull` here also includes anything which coerces to `null` in boolean context (i.e. any other type such as `array`).

With the following conversion from TypeNode:

- `boolean` => `{canBeTrue: true, canBeFalse: true, canBeNull: false}`
- `boolean | null` => `{canBeTrue: true, canBeFalse: true, canBeNull: true}`
- `unknown` => `{canBeTrue: true, canBeFalse: true, canBeNull: true}`

This at least makes it easier to reason about the logic.

E.g. for `And` we know that if either side is `canBeTrue: true, …: false` then it short-circuits and returns the other side. And if `canBeTrue: false, …` then everything becomes `false` actually.

TL;DR as I'm writing this: This `BooleanInterpretation` is maybe just a more precise version of `nodeMatch`.